spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala

```diff
@@ -17,80 +17,355 @@
 package org.apache.spark.sql.delta.perf
 
 import org.apache.spark.sql.catalyst.InternalRow
-import org.apache.spark.sql.catalyst.expressions.{Alias, Literal, ToPrettyString}
-import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, Complete, Count}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.planning.PhysicalOperation
 import org.apache.spark.sql.catalyst.plans.logical._
-import org.apache.spark.sql.delta.DeltaTable
+import org.apache.spark.sql.catalyst.util.{CaseInsensitiveMap, DateTimeUtils}
+import org.apache.spark.sql.delta.{DeltaColumnMapping, DeltaTable, Snapshot}
+import org.apache.spark.sql.delta.commands.DeletionVectorUtils.isTableDVFree
 import org.apache.spark.sql.delta.files.TahoeLogFileIndex
 import org.apache.spark.sql.delta.stats.DeltaScanGenerator
-import org.apache.spark.sql.functions.{coalesce, col, count, lit, sum, when}
-import org.apache.spark.sql.types.StringType
-import org.apache.spark.unsafe.types.UTF8String
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
 
+import java.sql.Date
+import java.util.Locale
+
+/** Optimize COUNT, MIN and MAX expressions on Delta tables.
+ * This optimization is only applied when the following conditions are met:
+ * - The MIN/MAX columns data type is supported by the optimization (ByteType, ShortType,
+ * IntegerType, LongType, FloatType, DoubleType, DateType).
+ * - All AddFiles in the Delta Log must have stats on columns used in MIN/MAX expressions,
+ * or the columns must be partitioned, in the latter case it uses partitionValues, a required field.
+ * - Table has no deletion vectors, or query has no MIN/MAX expressions.
+ * - COUNT has no DISTINCT.
+ * - Query has no filters.
+ * - Query has no GROUP BY.
+ * Example of valid query: SELECT COUNT(*), MIN(id), MAX(id) FROM MyDeltaTable
```

Comment:
Nit: Perhaps s/MAX(id)/MAX(partition_col), just to highlight that partition columns are stats-optimizable.